triton/stream.py

```diff
@@ -253,13 +260,227 @@ def checkpoint(self):
+            else:
+                self.first_next = False
+
+            if len(self.prefetch) == 0:
+                if len(self.pending_acks) > 0:
+                    self.subscription.acknowledge(self.pending_acks)
+                    self.pending_acks = []
+
+            pubsub_messages = self.subscription.pull(return_immediately=False)
+            for ack_id, raw_message in pubsub_messages:
+                self.prefetch_ack_ids.append(ack_id)
+                self.prefetch.append(self.stream.decode(raw_message))
+
+            if len(self.prefetch) == 0:
+                raise StopIteration
+
+            return self.prefetch.pop(0)
+
+
+    def __init__(self, project, topic, private_key_file):
+        self.http = None
+        self.credentials = None
+
+        # In order to support emulation we either pass
+        # user provided credentials into pubsub.Client or a
+        # simple http client.
+        if private_key_file:
+            self.credentials = service_account.Credentials.from_service_account_file(
+                private_key_file)
+        else:
+            self.http = requests.Session()
+
+        self.client = pubsub.Client(project, credentials=self.credentials, _http=self.http)
+        self.project = project
+        self.topic_name = topic
+        self.topic = self.client.topic(self.topic_name)
+
+        super(GCPStream, self).__init__()
+
+
+    @staticmethod
+    def retry_policy(e):
+        return True
+
+
+    def put(self, **record):
+        msg_id = _call_and_retry(
+            self.topic.publish,
+            self.encode(record),
+            retry_on_exception=self.retry_policy)
+
+        return (None, msg_id)
+
+
+    def put_many(self, records):
+        message_ids = []
+        packed = [self.encode(record) for record in records]
+
+        with self.topic.batch(max_messages = self.BATCH_MAX_MSGS) as topic_batch:
+            for pack in packed:
+                _call_and_retry(
+                    topic_batch.publish,
+                    pack,
+                    retry_on_exception=self.retry_policy
+                )
+
+            _call_and_retry(
+                topic_batch.commit,
+                retry_on_exception = self.retry_policy
+            )
+
+            # topic.batch defines an iterator
+            # over message_ids.
+            message_ids = list(topic_batch)
+
+        return [(None, message_ids)]
+
+
+    def decode(self, pubsub_message):
+        return super(GCPStream, self).decode(pubsub_message.data)
+
+
+    def build_iterator_from_latest(self):
+        return GCPStream.Iterator(self)
+
+
+    def build_iterator_from_checkpoint(self, subscription_id):
+        return GCPStream.Iterator(self, subscription_id)
+
+
+class AWSStream(Stream):
+
+    def __init__(self, name, partition_key, conn = None, region='us-east-1'):
+        if conn is not None:
+            self.conn = conn
+        else:
+            self.conn = connect_to_region(region)
```

Comment:
It's probably fine to just use `self.conn = conn or connect_to_region(region)` -- passing in other falsey things besides None is arguably a programming error though.